### DISCUSSION TITLE
Event Publisher in a Spring/JPA legacy application

### DISCUSSION THREAD

**Discussion by diegoicosta** [spring-projects/spring-modulith#769] [Q&A]
I'm very interested into how to apply Modulith in a legacy spring-boot app (3.3.1), specially the event publisher and event externalization.
I'm facing a issue where the JPA repository can't start because there are multiple primary EntityManager in the app. 
My question is, is there a way to avoid the auto-configuration and explicitly indicate a EntityManager qualifier to this JPA repository ?

```
Caused by: ... UnsatisfiedDependencyException: Error creating bean with name 'jpaEventPublicationRepository' defined in class path resource [org/springframework/modulith/events/jpa/JpaEventPublicationAutoConfiguration.class]: 

Unsatisfied dependency expressed through method 'jpaEventPublicationRepository' parameter 0: 

No qualifying bean of type 'jakarta.persistence.EntityManager' available: 
more than one 'primary' bean found among candidates: 
[ecomEntityManager, jpaSharedEM_AWC_ecomSessionFactory, jpaSharedEM_AWC_ecomEntityManagerFactory]
```

**Comment by diegoicosta** [comment]
@odrotbohm suggested to used JDBC, It worked

The `jpaSharedEM_AWC_` are created by Spring depending on the scope of transactions the applications is using, Its not explicitly define by the app. Eventually this can be something to think about as It can confuses devs while working with legacy apps with unclear transaction scope. 

Another relevant point: The notification log is persisted, but if the publisher method is not annotated with `@Transactional`, listeners are not triggered. I couldn't find any debug, error or warning in the logs about this. Pure try-and-error to figured out.

**Reply by odrotbohm** [reply]
Would you mind creating an issue, ideally with a reproducer attached? The EPR only interacts with `@TransactionalEventListener` to decide whether a publication log entry has to be created. Also note, that `@TEL` *have to* run with `@Transactional(Propagation.REQUIRES_NEW)` as they are executed in the cleanup phase of the original transaction. This is easy to overlook, and one of the reasons that`@ApplicationModuleListener` exists to default to that transactional setup in particular.

**Reply by gnchiesa** [reply]
Hi @diegoicosta @odrotbohm I'm so sorry to bother you guys with this, but I'm not following either of you. I have also multiple modules with their own datasoruces, but I'm hitting this issue. How can I make the event publishing to use JDBC ? I'm unable to access the JpaEventPublicationConfiguration directly; maybe I'm not understanding this correctly?
Annotating my "main" module dataSource and other configurations as `@Primary` makes everything work, but on shutdown I get a "Could not resolve root entity 'JpaEventPublication'". Could you shed some light on this, pretty please?

**Reply by diegoicosta** [reply]
I moved from JPA publisher to JDBC, and the issue to find a unique `EntityManager` was gone. The only change was switching the spring-boot starter dependency from: 
```
	 <dependency>
            <groupId>org.springframework.modulith</groupId>
            <artifactId>spring-modulith-starter-jpa</artifactId>
        </dependency>
```
to:
```
	 <dependency>
            <groupId>org.springframework.modulith</groupId>
            <artifactId>spring-modulith-starter-jdbc</artifactId>
        </dependency>
```
